orttraining/orttraining/python/training/utils/data/sampler.py

```diff
@@ -0,0 +1,340 @@
+        as the sample's computational complexity:
+        >>> dataset = torch.utils.data.TensorDataset(torch.randn(n, 2), torch.randperm(n))
+        >>> complexity_fn = lambda x: x[1]
+        Below is the usage of :class:`LoadBalancingDistributedSampler`
+        and `DataLoader <https://pytorch.org/docs/stable/data.html?highlight=dataloader#torch.utils.data.DataLoader>`_:
+        >>> sampler = bagua.torch_api.contrib.LoadBalancingDistributedSampler(
+        ...     dataset,
+        ...     complexity_fn=complexity_fn) if is_distributed else None
+        >>> loader = torch.utils.data.DataLoader(dataset,
+        ...     shuffle=(sampler is None),
+        ...     sampler=sampler)
+        >>>
+        >>> for epoch in range(start_epoch, n_epochs):
+        ...     if is_distributed:
+        ...         sampler.set_epoch(epoch)
+        ...     train(loader)
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        complexity_fn: Callable[..., int],
+        world_size: Optional[int] = None,
+        rank: Optional[int] = None,
+        shuffle: bool = True,
+        group_size: Optional[int] = None,
+        seed: int = 0,
+        drop_last: bool = False,
+        random_level: float = 0,
+    ) -> None:
+        if world_size is None:
+            if not dist.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            world_size = dist.get_world_size()
+        if rank is None:
+            if not dist.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            rank = dist.get_rank()
+        if rank >= world_size or rank < 0:
+            raise ValueError(
+                "Invalid rank {}, rank should be in the interval"
+                " [0, {}]".format(rank, world_size - 1)
+            )
+        self.dataset = dataset
+        self.world_size = world_size
+        self.rank = rank
+        self.epoch = 0
+        self.drop_last = drop_last
+        self.group_size = group_size
+
+        # If the dataset length is evenly divisible by number of shards, then there
+        # is no need to drop any data, since the dataset will be split equally.
+        dataset_len = len(self.dataset)
+        if self.drop_last and dataset_len % self.world_size != 0:
+            # Split to nearest available length that is evenly divisible.
+            # This is to ensure each rank receives the same amount of data when
+            # using this Sampler.
+            self.num_samples = dataset_len // self.world_size
+        else:
+            self.num_samples = math.ceil(dataset_len / self.world_size)
+        self.total_size = self.num_samples * self.world_size
+        self.shuffle = shuffle
+        self.seed = seed
+
+        self.sample_complexity_list = [None]*dataset_len
+        for sample_index in range(dataset_len):
```

Comment:
Why not moving that part in __iter__ or _sort_shard_and_shuffle_dataset? That would make the constructor a lot faster and the user would not pay a very long cost while building his dataset and his model.